rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+ - native support for all relevant OS's
+
+We will use the [go-piv](https://github.com/go-piv/piv-go) library, which is a Golang port of Yubikey's C library [ykpiv](https://github.com/Yubico/yubico-piv-tool/blob/master/lib/ykpiv.c). This is the same library used by [yubikey-agent](https://github.com/FiloSottile/yubikey-agent).
+
+#### PKCS#11
+
+PIV builds upon the more common cryptographical token interface ([PKCS#11](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/errata01/os/pkcs11-base-v2.40-errata01-os-complete.html#_Toc441755758)) which is used in cryptographical hardware devices, including HSMs, TPMs, and PIV-compatible smartcards. We could use PKCS#11 instead of PIV to support hardware-backed private keys across a broader range of hardware devices, but we would lose some key PIV features which we plan to make use of:
+ - PIV provides additional useful functionality, including generating new certificates, setting a key's PIN and other management data
+ - PIV provides a useful standard for PIN and touch requirements, which will give us a universal UX across PIV devices
+ - new hardware keys are likely to implement a PIV module (Such as the [solo2](https://solokeys.com/blogs/news/update-on-our-new-and-upcoming-security-keys))
+
+If we want to expand support for non-PIV cryptographical devices in the future, we could add this functionality separately. TPM support in particular could be used to broaden hardware-backed key support to a wide array of modern computers, including all Windows 11 systems.
+
+#### Non-yubikey device compatibility
+
+Currently, Yubikey is one of the only PIV-compatible commercial hardware keys. As a result, current PIV implementations are specifically designed around Yubikey's implementation of PIV - the `libykcs11.so` module. While the majority of PIV is standardized, `libykcs11.so` has [some extensions](https://developers.yubico.com/PIV/Introduction/Yubico_extensions.html) and other idiosyncracies which may not be standard accross future PIV implemenations.
+
+Unfortunately, there is no common PIV library, so our best option is to use `piv-go` for a streamlined implemenation and prepare to adjust in the future as more PIV-compatible hardware keys are released. Possible adjustments include:
+ - using multiple PIV libraries to support custom PIV implementations
+ - switching to a PIV library which expressly supports all/more PIV implemenations
+ - working within a PIV library, through PRs or a Fork, to expand PIV support
+ - creating our own custom PIV library which we can add custom support into as needed
+
+Note: the adjustments above will largely be client-side and therefore should not pose any backwards compatibility concerns.
+
+### tsh PIV integration
+
+We'll add the `--piv` flag so that users can use `tsh login --piv` to login with their PIV card. The majority of the login flow is the same, except the key generation and certificate signing process.
```

Comment:
We should make sure this works with tsh aliases for folks that want to make this default behavior. Should just work, just something to test.

Reply:
Also, I think I asked before but if PIV usage is enforced by the server, can users just omit this flag?

Reply:
For the second comment - yes, I've added a section for this here: https://github.com/gravitational/teleport/blob/6026bb92270ed6e523acdd27282900a427c69c10/rfd/0080-yubikey-piv.md#retrieve-users-piv-attestation-requirements-during-login